platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractCollectionProperty.java

```diff
@@ -574,55 +689,40 @@ private ExecutionTimeValue<C> getFixedExecutionTimeValue(List<ExecutionTimeValue
 
         @Override
         public ValueProducer getProducer() {
-            return value.getProducer();
-        }
-
-        @Override
-        public String toString() {
-            return value.toString();
-        }
-    }
-
-    /**
-     * A provider for a collection type whose elements are themselves providers.
-     */
-    private static class CollectingProvider<T, C extends Collection<? extends T>> extends AbstractMinimalProvider<C> {
-        private final Class<C> type;
-        private final List<ProviderInternal<? extends Iterable<? extends T>>> providers;
-        private final Supplier<ImmutableCollection.Builder<T>> collectionFactory;
+            List<ValueProducer> producers = getCollectors().stream().map(ValueSupplier::getProducer).collect(toList());
+            return new ValueProducer() {
+                @Override
+                public void visitProducerTasks(Action<? super Task> visitor) {
+                    producers.forEach(c -> c.visitProducerTasks(visitor));
+                }
 
-        public CollectingProvider(Class<C> type, List<ProviderInternal<? extends Iterable<? extends T>>> providers, Supplier<ImmutableCollection.Builder<T>> collectionFactory) {
-            this.type = type;
-            this.providers = providers;
-            this.collectionFactory = collectionFactory;
-        }
+                @Override
+                public boolean isKnown() {
+                    return producers.stream().anyMatch(ValueProducer::isKnown);
+                }
 
-        @Nullable
-        @Override
-        public Class<C> getType() {
-            return type;
-        }
+                @Override
+                public void visitDependencies(TaskDependencyResolveContext context) {
+                    producers.forEach(c -> c.visitDependencies(context));
+                }
 
-        @Override
-        public ExecutionTimeValue<? extends C> calculateExecutionTimeValue() {
-            return ExecutionTimeValue.changingValue(this);
+                @Override
+                public void visitContentProducerTasks(Action<? super Task> visitor) {
+                    producers.forEach(c -> c.visitContentProducerTasks(visitor));
+                }
+            };
         }
 
         @Override
-        protected Value<? extends C> calculateOwnValue(ValueConsumer consumer) {
-            ImmutableCollection.Builder<T> builder = collectionFactory.get();
-            SideEffectBuilder<? super C> sideEffectBuilder = SideEffect.builder();
-            for (ProviderInternal<? extends Iterable<? extends T>> provider : providers) {
-                Value<? extends Iterable<? extends T>> value = provider.calculateValue(consumer);
-                if (value.isMissing()) {
-                    return Value.missing();
+        protected String toStringNoReentrance() {
+            StringBuilder sb = new StringBuilder();
+            getCollectors().forEach(collector -> {
+                if (sb.length() > 0) {
```

Comment:
🚀 Consider limiting the size of the produced string.

Reply:
> 🚀 Consider limiting the size of the produced string.

I'd leave it out of this PR. It was working like that before, though hidden in the recursive calls.

Reply:
> I'd leave it out of this PR. 

Ok.